Source/ExecutionEngine/ExecutionEngine.cs

```diff
@@ -11,174 +11,16 @@
-  {
-    public virtual ErrorInformation CreateErrorInformation(IToken tok, string msg, string requestId = null,
-      string originalRequestId = null, string category = null)
-    {
-      Contract.Requires(1 <= tok.line && 1 <= tok.col);
-      Contract.Requires(msg != null);
-
-      return ErrorInformation.CreateErrorInformation(tok, msg, requestId, originalRequestId, category);
-    }
-  }
-
-
-  public class ErrorInformation
-  {
-    public readonly IToken Tok;
-    public string Msg;
-    public string Category { get; set; }
-    public readonly List<AuxErrorInfo> Aux = new List<AuxErrorInfo>();
-    public string OriginalRequestId { get; set; }
-    public string RequestId { get; set; }
-    public ErrorKind Kind { get; set; }
-    public string ImplementationName { get; set; }
-    public TextWriter Out = new StringWriter();
-    public TextWriter ModelWriter = new StringWriter();
-
-    public string FullMsg
-    {
-      get
-      {
-        return Category != null ? string.Format("{0}: {1}", Category, Msg) : Msg;
-      }
-    }
-
-    public struct AuxErrorInfo
-    {
-      public readonly IToken Tok;
-      public readonly string Msg;
-      public readonly string Category;
-
-      public string FullMsg
-      {
-        get { return Category != null ? string.Format("{0}: {1}", Category, Msg) : Msg; }
-      }
-
-      public AuxErrorInfo(IToken tok, string msg, string category = null)
-      {
-        Tok = tok;
-        Msg = CleanUp(msg);
-        Category = category;
-      }
-    }
-
-    protected ErrorInformation(IToken tok, string msg)
-    {
-      Contract.Requires(tok != null);
-      Contract.Requires(1 <= tok.line && 1 <= tok.col);
-      Contract.Requires(msg != null);
-
-      Tok = tok;
-      Msg = CleanUp(msg);
-    }
-
-    internal static ErrorInformation CreateErrorInformation(IToken tok, string msg, string requestId = null,
-      string originalRequestId = null, string category = null)
-    {
-      var result = new ErrorInformation(tok, msg);
-      result.RequestId = requestId;
-      result.OriginalRequestId = originalRequestId;
-      result.Category = category;
-      return result;
-    }
-
-    public virtual void AddAuxInfo(IToken tok, string msg, string category = null)
-    {
-      Contract.Requires(tok != null);
-      Contract.Requires(1 <= tok.line && 1 <= tok.col);
-      Contract.Requires(msg != null);
-      Aux.Add(new AuxErrorInfo(tok, msg, category));
-    }
-
-    protected static string CleanUp(string msg)
-    {
-      if (msg.ToLower().StartsWith("error: "))
-      {
-        return msg.Substring(7);
-      }
-      else
-      {
-        return msg;
-      }
-    }
-  }
-
-  #endregion
-
   public class ExecutionEngine : IDisposable {
-    public static ErrorInformationFactory ErrorInformationFactory { get; } = new();
+
+    private static readonly ConditionalWeakTable<Program, Action<VCGen, Implementation, VerificationResult>> ResultPostProcessors = new ();
```

Comment:
I understand that you are setting a callback for when the Boogie program is done verifying, so that the routines of `ExtractLoop` can be performed independently of the verification engine and these callbacks that don't prevent the garbage collector to be collected if no reference points to Program anymore (e.g. when verification is canceled)

I also understand that you want to run these action synchronously, so you are not using TaskCompletionSource.

I'm puzzled because you are not deleting the callback when it is processed, which I think it should. I'm not sure about garbage collection, but for example in Java it's lazy and might not follow the links of a ConditionalWeakTable to determine whether garbage collection is needed, but programs can be quite heavy.

* The principle of "hiding away" this logic is hindered by the fact that there is also at most one Action to run per program, so it's not even an architecture that will enable more use cases.
* Could you perhaps have a synchronously executed task instead?
* I would prefer to need not a static shared variable caching all these callbacks, but rather somehow store the callback in the execution that needs it, so that there is no need for a table.

Please let me know what you think of the above.

Reply:
I've replaced the use of `ConditionalWeakTable` with an extra record. Curious if you like the new implementation better.

Reply:
That's great ! I was wondering how you would do the change, and what you did made me better realize the flow, i.e. where is this callback is being used. At least for me it's definitely more readable that way, compared to the previous "stateful" implementation. I thank you for the effort!
I'll have a few suggestions about the naming conventions.